tests/regression/test_restricted_function_space.py

```diff
@@ -0,0 +1,245 @@
+import pytest
+import numpy as np
+from firedrake import *
+
+
+def compare_function_space_assembly(function_space, restricted_function_space,
+                                    bcs, res_bcs=[]):
+    u = TrialFunction(function_space)
+    v = TestFunction(function_space)
+    original_form = u * v * dx
```

Comment:
```suggestion
    original_form = inner(u, v) * dx
```
to make it complex safe. The same change must happen everywhere in the below.